misc/python/materialize/data_ingest/executor.py

```diff
@@ -322,3 +334,133 @@ def run(self, transaction: Transaction) -> None:
+                    ENVELOPE DEBEZIUM""",
+            )
+            self.execute(
+                cur,
+                f"""CREATE SOURCE {self.table}
+                    FROM KAFKA CONNECTION kafka_conn (TOPIC '{self.topic}')
+                    FORMAT AVRO
+                    USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn
+                    ENVELOPE DEBEZIUM""",
+            )
+        self.mz_conn.autocommit = False
+
+    def run(self, transaction: Transaction) -> None:
+        with self.mz_conn.cursor() as cur:
+            for row_list in transaction.row_lists:
+                for row in row_list.rows:
+                    key_values = tuple(
+                        value
+                        for field, value in zip(row.fields, row.values)
+                        if field.is_key
+                    )
+                    if row.operation == Operation.INSERT:
+                        values_str = ", ".join(
+                            str(formatted_value(value)) for value in row.values
+                        )
+                        self.execute(
+                            cur,
+                            f"""INSERT INTO {self.table_original}
+                                VALUES ({values_str})
+                            """,
+                        )
+                        self.known_keys.add(key_values)
+                    elif row.operation == Operation.UPSERT:
+                        if key_values in self.known_keys:
+                            non_key_values = tuple(
+                                (field, value)
+                                for field, value in zip(row.fields, row.values)
+                                if not field.is_key
+                            )
+                            # Can't update anything if there are no values, only a key, and the key is already in the table
+                            if non_key_values:
+                                cond_str = " AND ".join(
+                                    f"{field.name} = {formatted_value(value)}"
+                                    for field, value in zip(row.fields, row.values)
+                                    if field.is_key
+                                )
+                                set_str = ", ".join(
+                                    f"{field.name} = {formatted_value(value)}"
+                                    for field, value in non_key_values
+                                )
+                                self.mz_conn.autocommit = True
+                                self.execute(
+                                    cur,
+                                    f"""UPDATE {self.table_original}
+                                        SET {set_str}
+                                        WHERE {cond_str}
+                                    """,
+                                )
+                                self.mz_conn.autocommit = False
```

Comment:
Are we changing the autocommit behavior just for this one `UPDATE` statement? If so, what about using an explicit commit instead?

Reply:
Explicit commit doesn't work unfortunately, as soon as you have `BEGIN; UPDATE` you get the error already, so before you have the chance to commit. Can be reproduced in psql too.